doc/rfc-ephemeral-disk-service.md

```diff
@@ -0,0 +1,466 @@
+  - label: local1
+    filesystem: ext4
+    device: /dev/disk/azure/local/by-index/1
+
+mounts:
+  - [ "/dev/disk/azure/local/by-index/1", "/mnt", "ext4", "defaults,nofail,x-systemd.makefs", "0", "2" ]
+```
+
+In this example, `x-systemd.makefs` is used to ensure systemd formats the disk if VM is reallocated.
+
+# Design
+
+## Goals
+
+The goals for the design are as follows:
+
+* Automatic detection, formatting and mounting of ephemeral disks.
+* Aggregate multiple local NVMe disks with mdadm using RAID-0.
+* Support configurable filesystem types and mount points.
+* Configurable knobs to gracefully handle potential conflicts with `cloud-init` and `WALinuxAgent`.
+* Minimize boot performance impact.
+* Reliable booting and graceful handling of stale or incorrect configurations.
+
+## Service Naming
+
+Open to suggestion on names.  Keeping open the possibility of handling SCSI resource disks:
+
+* azure-ephemeral-disk-setup.service
+* azure-ephemeral-disks.service
+
+## Service Ordering
+
+Considerations:
+
+* We want order ourselves *before* `cloud-init.service` so we can avoid race conditions in setup of the SCSI resource disk.  It also may give us the opportunity to disable Azure resource disk formatting in cloud-init if it's available in scenarios (b), (c), and (d).
+* We would typically desire `RequiredMountsFor=$AZURE_EPHEMERAL_DISK_SETUP_MOUNT_POINT` but we can't do this dynamically.
+* cloud-init-local.service only runs `After=systemd-remount-fs.service`. If we put a dependency of `Before=cloud-init-local.service` and `After=local-fs.target` we may slightly impact boot performance.
+* We can likely avoid a potential `udevadm settle` for typical boot if we order ourselves after `local-fs.target` and check for presence of mount at configured mount point.
+
+Proposed `After=`:
+
+* local-fs.target: this mitigates concerns with `RequiresMountsFor=$AZURE_EPHEMERAL_DISK_SETUP_MOUNT_POINT` and allows us to check if something is already mounted at the configured mount point.
+
+Proposed `Before=`:
+
+* cloud-init.service: avoids race condition with configuration/setup of disks
+* sysinit.target: reasonable target to ensure most services likely to be dependent on mount will start after
+
+**Concerns:**
+
+* This may be some additional delay with the ordering of `local-fs.target -> azure-ephemeral-disk-setup.service -> cloud-init.service -> sysinit.target`.  We'll want to do some measurements of this.
+
+## Systemd Service
+
+Unit: `azure-ephemeral-disk-setup.service`
+
+```ini
+[Unit]
+Description=Format and mount local ephemeral NVMe disks
+After=local-fs.target
+Before=cloud-init.service sysinit.target
+
+[Service]
+Type=oneshot
+EnvironmentFile=/etc/default/azure-ephemeral-disk-setup
+ExecStart=/usr/sbin/azure-ephemeral-disk-setup.sh
+
+[Install]
+WantedBy=sysinit.target
+```
+
+## Service Configuration
+
+To allow for some flexibility for customers, we initially intend to support a number of environment variables to control behaviors:
+
```

Comment:
here is one question about use env variables to control behaviors. pls correct me if I misunderstand.

currently, if we use marketplace image, the common way to set environment variables in azure VMSS is to pass cloud-init configuration as custom data. If the `azure-ephemeral-disk-setup.service` start before `cloud-init.service`, maybe it   is hard to pass environment variables to VM.